Host variation data practical session
=====================================
.. hint::
    This practical session uses software available on the course-provided virtual machines.
    To follow this workshop at a later date, see `the github repo <https://github.com/ebi-metagenomics/holofood-course>`_ for installation instructions.

microbiome-GWAS using GEMMA
---------------------------

1. Prepare microbiome composition data
2. Prepare individual covariate data
3. Run GEMMA
4. Visualise GWAS results
5. Extract SNP annotation

Prerequisites
~~~~~~~~~~~~~

For this tutorial you will need to first load the conda environment by running:

.. code-block:: 
	
	conda activate mgwas-env

The software GEMMA is already installed on the virtual machine. The user manual can be found on the `GEMMA github repo <https://github.com/genetics-statistics/GEMMA>`_. 

The rest of this practical is available on a `dedicated page </_static/host-variation-practical.html>`_ (which is also downloadable from `the github repo <https://github.com/ebi-metagenomics/holofood-course/>`_).
